### PR TITLE
Adds -lm linker flag for FreeBSD/OpenBSD and uses github.com/bazelbuil…

### PR DESCRIPTION
**Pull request by yesudeep**
…d/platforms for platform detection.

Addresses: https://github.com/google/benchmark/issues/1031

### CODE REVIEW

**Comment by yesudeep** [comment]
This also addresses a broken build for google/cctz on FreeBSD.

```shell
ERROR: /usr/home/yesudeep/code/yesudeep/cctz/BUILD:120:8: Linking of rule '//:cctz_benchmark' failed (Exit 1) clang failed: error executing command /usr/bin/clang -o bazel-out/freebsd-fastbuild/bin/cctz_benchmark bazel-out/freebsd-fastbuild/bin/_objs/cctz_benchmark/cctz_benchmark.pic.o bazel-out/freebsd-fastbuild/bin/libtime_zone.a ... (remaining 8 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
ld: error: undefined symbol: ceilf
>>> referenced by time_zone_impl.cc
>>>               time_zone_impl.pic.o:(ceil(float)) in archive bazel-out/freebsd-fastbuild/bin/libtime_zone.a

ld: error: undefined symbol: lround
>>> referenced by benchmark_runner.cc
>>>               benchmark_runner.pic.o:(benchmark::internal::(anonymous namespace)::BenchmarkRunner::PredictNumItersNeeded(benchmark::internal::(anonymous namespace)::BenchmarkRunner::IterationResults const&) const) in archive bazel-out/freebsd-fastbuild/bin/external/com_github_google_benchmark/libbenchmark.a
>>> referenced by json_reporter.cc
>>>               json_reporter.pic.o:(benchmark::(anonymous namespace)::RoundDouble(double)) in archive bazel-out/freebsd-fastbuild/bin/external/com_github_google_benchmark/libbenchmark.a

ld: error: undefined symbol: log
>>> referenced by complexity.cc
>>>               complexity.pic.o:(benchmark::FittingCurve(benchmark::BigO)::$_3::operator()(unsigned long) const) in archive bazel-out/freebsd-fastbuild/bin/external/com_github_google_benchmark/libbenchmark.a
>>> referenced by complexity.cc
>>>               complexity.pic.o:(benchmark::FittingCurve(benchmark::BigO)::$_4::operator()(unsigned long) const) in archive bazel-out/freebsd-fastbuild/bin/external/com_github_google_benchmark/libbenchmark.a

ld: error: undefined symbol: pow
>>> referenced by complexity.cc
>>>               complexity.pic.o:(std::__1::_MetaBase<(std::is_arithmetic<double>::value) && (std::is_arithmetic<int>::value)>::_EnableIfImpl<std::__1::__promote<double, int, void> >::type pow<double, int>(double, int)) in archive bazel-out/freebsd-fastbuild/bin/external/com_github_google_benchmark/libbenchmark.a
>>> referenced by complexity.cc
>>>               complexity.pic.o:(std::__1::_MetaBase<(std::is_arithmetic<unsigned long>::value) && (std::is_arithmetic<int>::value)>::_EnableIfImpl<std::__1::__promote<unsigned long, int, void> >::type pow<unsigned long, int>(unsigned long, int)) in archive bazel-out/freebsd-fastbuild/bin/external/com_github_google_benchmark/libbenchmark.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)
INFO: Elapsed time: 0.261s, Critical Path: 0.10s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

File: BUILD.bazel [hunk 1]
```diff
@@ -22,6 +38,14 @@ cc_library(
     hdrs = ["include/benchmark/benchmark.h"],
     linkopts = select({
         ":windows": ["-DEFAULTLIB:shlwapi.lib"],
+        ":freebsd": [
```
Comment (dominichamon) [inline BUILD.bazel:41]: if we `load("@bazel_skylib//lib:selects.bzl", "selects")` then we can replace this with:

```
linkopts = selects.with_or({
    ":windows": ["-DEFAULTLIB:shlwapi.lib"],
    (":freebsd", ":openbsd"): ["-pthread", "-lm"],
})
```

which i think is going to be easier to read/maintain. assuming freebsd/openbsd don't diverge.

as an alternative, you can define:

```
selects.config_setting_group(
    name = "bsd",
    match_any = [":freebsd", ":openbsd"],
)
```

and then just use ":bsd" as the key here.

up to you :)

Comment (yesudeep) [reply BUILD.bazel]: That would work within the repository, but then it requires every dependent repository to install @bazel_skylib too, I guess. It turns out that Bazel doesn't load `WORKSPACE` dependencies transitively (?), and the "@platforms" package appears to load implicitly. I'm wondering whether these Bazel utility repositories should be built-ins or be loaded by default.

I've done something simpler: defined a variable named `BSD_LINKOPTS` and shared it with both the ":freebsd" and ":openbsd" keys. Would that work for you?

Comment (yesudeep) [reply BUILD.bazel]: I've filed an [issue](https://github.com/bazelbuild/bazel/issues/12014) with the Bazel team. I like the approach you've suggested, but adding a dependency here would complicate things for users, which I'm not so sure they'd like.

Comment (dominichamon) [reply BUILD.bazel]: that's a great compromise. thanks!

**Pull request merged**